group_vars/all/vars.yml

```diff
@@ -20,6 +30,15 @@ public_keys_of_local_admins: "{{ local_admin_users | map('extract', auth_users)
     | map(attribute='pub_keys') \
     | map('regex_findall', '^([^\\n]+)', 'multiline=True') | flatten }}"
 #
+# Networking.
+#
+host_network_names: "{{ host_networks | map(attribute='name') | flatten | unique }}"
+first_management_network_name: "{{ stack_networks
```

Comment:
If I understand correctly this picks the "first" management network. Which can be anything from the list. Shouldn't we have variable per network, f.e. `main: true` or `first: true` and select that only one? And if none was defined as such, the error would be also helpful.

Or did I misunderstood?

Reply:
You understood correctly. It simply picks "a" management network from the list when there are multiple: 'first' depends on the order/sorting; the code could have used `last` instead and the effect would be the same. The IPs from `first_management_network_name` are used in `/etc/hosts` deployed on machines and hence will be used to access a machine based on its hostname. This currently assumes all "management networks" are equally suitable for this purpose. Note this currently only happens on the iRODS machines, which have 2 management networks. If a specific network should be used to create entries in `/etc/hosts`, then we need an additional attribute/value to select a specific one. The alternative is to introduce a different category. Maybe something like _primary management network_ of which there can be only one and adding additional management networks to a list of _secondary management networks_. Or ...

Reply:
> This currently assumes all "management networks" are equally suitable for this purpose.
But if the list of networks changes, the entire playbook should be rerun on all the machines, to change all the network settings.

First network today could not be first network tomorrow. Or last one - doesn't matter which one we look.

Everything can be reshuffled simply by adding another VLAN, as has happened before :wink: 

> If a specific network should be used to create entries in /etc/hosts, then we need an additional attribute/value to select a specific one.
No, I like the fact that the `/etc/hosts` has all the machines from all the networks defined. This is just a phone book and I prefer that it knows all the citizens. It has been useful before.

Reply:
> No, I like the fact that the /etc/hosts has all the machines from all the networks defined. This is just a phone book and I prefer that it knows all the citizens. It has been useful before.

Yes, you can list all the machines, but you cannot list all the interfaces of a machine and give them the same name. Let's assume a machine by the name of `supernode` has 2 interfaces in an internal "management" network:

- 10.0.1.1 in management VLAN1 with subnet 10.0.1.0/24
- 10.0.2.1 in management VLAN2 with subnet 10.0.2.0/24

You can either put in `/etc/hosts`:
```
10.0.1.1     supernode
```
or you can put in `/etc/hosts`:
```
10.0.2.1     supernode
```
If both mangement networks are equally suitable to access the machine it does not matter, which one is used. But you cannot list both like this without creating all kinds of network issues:
```
10.0.1.1     supernode
10.0.2.1     supernode
```
If you want to list both interfaces in `/etc/hosts` only one of them can have the hostname of the machine. The other one would need another name. Something like this could work:
```
10.0.1.1     supernode
10.0.2.1     supernode_in_vlan2
```
An IP can have multiple names and aliases, so this would also work:
```
10.0.1.1     supernode_in_vlan1    supernode
10.0.2.1     supernode_in_vlan2
```
If you want to assign the name `supernode` to a specific IP, we need an additional attribute or variable to select a specific IP for `/etc/hosts`.